core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the “License”). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.file;
+
+import alluxio.AbstractClient;
+import alluxio.AlluxioURI;
+import alluxio.Configuration;
+import alluxio.Constants;
+
+import alluxio.client.file.options.CancelUfsFileOptions;
+import alluxio.client.file.options.CloseUfsFileOptions;
+import alluxio.client.file.options.CompleteUfsFileOptions;
+import alluxio.client.file.options.CreateUfsFileOptions;
+import alluxio.client.file.options.OpenUfsFileOptions;
+import alluxio.exception.AlluxioException;
+import alluxio.exception.ConnectionFailedException;
+import alluxio.heartbeat.HeartbeatContext;
+import alluxio.heartbeat.HeartbeatExecutor;
+import alluxio.heartbeat.HeartbeatThread;
+import alluxio.thrift.AlluxioService;
+import alluxio.thrift.AlluxioTException;
+import alluxio.thrift.FileSystemWorkerClientService;
+import alluxio.util.network.NetworkAddressUtils;
+import alluxio.wire.WorkerNetAddress;
+import alluxio.worker.ClientMetrics;
+
+import com.google.common.base.Preconditions;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+/**
+ * Client for talking to a file system worker server. It keeps sending keep alive messages to the
+ * worker server to preserve its state.
+ *
+ * Since {@link alluxio.thrift.FileSystemWorkerClientService} is not thread safe, this class
+ * guarantees thread safety.
+ */
+// TODO(calvin): Session logic can be abstracted
+@ThreadSafe
+public class FileSystemWorkerClient extends AbstractClient {
```

Comment:
sg, but I think it might worth adding some unit tests in the long run. It's very critical logic path. Maybe create a JIRA ticket as TODO?